src/vm/docs/mod.rs

```diff
@@ -815,6 +815,21 @@ supplied), this function returns `none`.
 "#,
 };
 
+const SLICE_API: SpecialAPI = SpecialAPI {
+    input_type: "sequence_A, uint, uint",
+    output_type: "(optional sequence_A)",
+    signature: "(slice sequence position length)",
+    description:
+        "The `slice` function attempts to return a sub-sequence of size `length` starting from
+index `position` in the provided sequence.
+If `length` is 0 or `position + length` is greater than `(len sequence)`, this function returns `none`.",
```

Comment:
maybe there is some variation in languages, but it's more usual to return a partial list if `length` will overflow the original sequence.

also, if `position > length` I would return an empty list instead of "none"... where did you get this API from?

Also, why are we using "length" rather than "left_index"/"right_index"?

This is what happens in python, for example:

```
>>> l=[1,2,3,4,5]
>>> l[10:]
[]
>>> l[3:10]
[4, 5]
```

Reply:
Also I'm not too attached to using position/length over left_index/right_index, we can see what others think

Reply:
I'm also a fan of returning a result/option when the slice indices are out of range. I think this is more in the spirit of the Clarity lang, and is consistent with how functions like `element-at` behave

Reply:
I guess in Rust it just panics if you try to use an inappropriate range:

```
fn main() {
    // Statements here are executed when the compiled binary is called

    // Print text to the console
    println!("Hello World!");

		let a = vec!(1,2,3,4);
    println!("a {:?}", a);

		let b = &a[1..3];
    println!("b {:?}", b);

		// Panics.
		let c = &a[1..10];
    println!("c {:?}", c);
}
```

c++ seems to use undefined behavior as well:
https://en.cppreference.com/w/cpp/container/span/span

Reply:
That being said, I think that if `length` is `0`, an empty list should be returned, rather than `none`. I think it's still best to return `none` if `position + length > (len sequence)` though.

Reply:
I'm also curious about why you decided to use `position/length` instead of indices.

I think that returning an empty list is better for when the slice `length` is `0`, since the parameters are not technically invalid in the way that they are when you provide a length that puts you out of bounds.

Reply:
Don't have an opinion about `position/length` vs indices

Reply:
+1 on the current approach, I think that taking a Rust-like approach is more in line with Clarity's design, than Python.

Reply:
1) If, as Ludo says, we are being Rust style, that would mean:
a) Fail on invalid indices. Why not just return an error for bad indices?
b) Use start/end, e.g. `let b = &a[1..3]` in the example I posted has value `b [2, 3]`. It seems like we didn't really make a clear decision on this.

2) As Reed says, I think list of length 0 should be an empty list, not a None.

Also please include the 0 length sub-sequence as an example here.

Reply:
@kantai @jcnelson any opinions here?

Reply:
I'm a pretty firm believer that we should stick to a rust-like error style here -- that is, providing an out of bounds index should result in a `none` result.

However, I do agree with @reedrosenbluth for slice length 0 -- we should just return an empty list rather than `none`.

Reply:
@kantai Thanks for your input! Any opinions on index/length vs. left/right indices?

Reply:
Yep -- I prefer left-right indexes, rather than a length arg.